### PR TITLE
feat(data-table): use iterate differ to optimize rendering

### PR DESCRIPTION
**Pull request by andrewseguin**

### CODE REVIEW

File: src/lib/core/data-table/data-table.spec.ts [hunk 1]
```diff
@@ -97,6 +96,39 @@ describe('CdkTable', () => {
...
+
+    // Prove that the attributes match their indicies
+    const initialRows = getRows(tableElement);
+    expect(initialRows[0].getAttribute('initialIndex')).toBe('0');
+    expect(initialRows[1].getAttribute('initialIndex')).toBe('1');
+    expect(initialRows[2].getAttribute('initialIndex')).toBe('2');
+
+    // Swap first and second data in data array
+    let copiedData = component.dataSource.data.slice();
+    let temp = copiedData[0];
```
Comment (ErinCoughlan) [inline src/lib/core/data-table/data-table.spec.ts]: let -> const in both of these

File: src/lib/core/data-table/data-table.ts [hunk 2]
```diff
@@ -92,9 +98,13 @@ export class CdkTable implements CollectionViewer {
   /** Set of templates that used as the data row containers. */
   @ContentChildren(CdkRowDef) _rowDefinitions: QueryList<CdkRowDef>;
 
-  constructor(private _changeDetectorRef: ChangeDetectorRef) {
+  constructor(private _differs: IterableDiffers,
+              private _changeDetectorRef: ChangeDetectorRef) {
```
Comment (ErinCoughlan) [inline src/lib/core/data-table/data-table.ts]: Should both of these be readonly as well?

Comment (andrewseguin) [inline src/lib/core/data-table/data-table.ts]: I don't think we've set that precedent, but I can talk with Jeremy about adding it to here and other components as well

Comment (jelbourn) [inline src/lib/core/data-table/data-table.ts]: Yeah, they should be `readonly`; we started the library before `readonly` was a thing and never got in the habit of adding it later

File: src/lib/core/data-table/data-table.ts [hunk 3]
```diff
@@ -76,6 +79,9 @@ export class CdkTable implements CollectionViewer {
    */
   private _columnDefinitionsByName = new Map<string,  CdkColumnDef>();
 
+  /** Differ used to find the changes in the data provided by the data source. */
+  private _dataDiffer: IterableDiffer<any> = null;
```
Comment (jelbourn) [inline src/lib/core/data-table/data-table.ts]: One thing we should consider doing before this is released is making it `CdkTable<T>` and using the generic throughout

File: src/lib/core/data-table/data-table.ts [hunk 4]
```diff
@@ -92,9 +98,13 @@ export class CdkTable implements CollectionViewer {
...
   @ContentChildren(CdkRowDef) _rowDefinitions: QueryList<CdkRowDef>;
 
-  constructor(private _changeDetectorRef: ChangeDetectorRef) {
+  constructor(private _differs: IterableDiffers,
+              private _changeDetectorRef: ChangeDetectorRef) {
     console.warn('The data table is still in active development ' +
         'and should be considered unstable.');
+
+    // TODO(andrewseguin): Add trackby function input.
+    this._dataDiffer = this._differs.find([]).create();
```
Comment (jelbourn) [inline src/lib/core/data-table/data-table.ts]: I don't quite follow how this works; why is this `find` necessary? Can you not just create an `IterableDiffer` directly? Why is the DI system involved? 
(probably just needs explanation in comments)

File: src/lib/core/data-table/data-table.ts (hunk 3 above)
Comment (andrewseguin) [inline src/lib/core/data-table/data-table.ts]: Added a <T> generic to the CdkTable definition

File: src/lib/core/data-table/data-table.ts (hunk 4 above)
Comment (andrewseguin) [inline src/lib/core/data-table/data-table.ts]: Added some comment mentioning that we need to find an iterable differ used for array-based iterables. It's the only public API that lets us grab it, no luck instantiating one myself.

**Comment by jelbourn** [comment]
Just noticed still needs `readonly` per Erin's comment

**Comment by andrewseguin** [comment]
Thanks for catching that, got the readonly in, thought I had changed that already.

**Pull request merged**